### PR TITLE
[docker] add basic Dockerfile

### PR DESCRIPTION
**Pull request by tmc**

### CODE REVIEW

**Comment by okayzed** [comment]
i'm not familiar with docker - how do you test this and did it work? should the dockerfile also include the ability to bundle sybil too? 

there's another repo that i've been using for building AppImage and Snap packages for snorkel+sybil: https://github.com/logv/snapcraft. snapcraft has a read-only package + a separate writeable dir, so i had to make changes to how snorkel works to accommodate that. not sure if docker has similar requirements.

**Comment by prologic** [comment]
No; sybil should be a separate image.

**Comment by okayzed** [comment]
i'm building it now - i also see where you got the npm audit log output. 

despite not knowing docker (or its conventions), i have a Dockerfile I started for sybil, but didn't get far with:  https://github.com/logv/sybil/tree/docker/docker, maybe it can be useful for you.

**Comment by prologic** [comment]
I've been following this project for a few weeks. I'll help with Dockerizing this :)

**Comment by okayzed** [comment]
@tmc: when i run the docker image locally (after doing a `docker build`), it errors with a missing htpasswd node module, but i think snorkel's npm-shrinkwrap lists htpasswd as a dep (maybe there is some misconfiguration in snorkel's npm shrinkwrap?)

@prologic perfect, thanks!

**Comment by tmc** [comment]
@okayzed sorry was missing the shrinkwrap file when I added separate cache utilization of package.json, this has been updated and tested.

**Comment by okayzed** [comment]
the image builds and i can run it. but there are no users, so i can't log in yet and test more stuff out. i have to look into where docker will store the app data and whether docker has the concept of read-only container + write area for user data. i'm also curious to see how sybil will get tied in

maybe some more stuff to do:

* add entry point (or other method) for adding users (`scripts/add_user.sh <username>`). users are kept in htpasswd file in `config/users.htpasswd` and is specified in the config.js. at the moment there is no UX for adding users through the snorkel UI. 

* add `ENV=docker` to the app invocation with docker specific settings. can see how [snapcraft config](https://github.com/logv/snorkel/blob/master/snorkel/config/snapcraft.js) is setup.  to enable snapcraft (or other) config, use `ENV=snapcraft node app.js`. the snapcraft config is setup s.t. all config and data is being written into a separate directory from the actual code (data_dir ends up being ~/snap/snorkel/common) because the snapcraft package (all the snorkel code) is read-only

**Comment by tmc** [comment]
@okayzed I'll introduce a docker.js config. -- user management scripts can be reached via docker exec invocations. While sybil certainly should have it's own docker image(s) I think it'd be most appropriate for you do publish a binary release of sybil that can be pulled into the snorkel image.

**Comment by tmc** [comment]
@okayzed PTAL

**Comment by prologic** [comment]
I would really not be doing this at all. Don't expect to pull in sybil into the snorkel image. Keep things decoupled and keep "scalability" in mind. Don't back yourself into a corner like this; build the images seaprately and think about "scale".

**Comment by okayzed** [comment]
in terms of scaling, what i was picturing (based on scuba design) for distributed queries is something like:

* leaf nodes: many machines will have a sybil install, perhaps connected to the end of ingestion pipes (like scribe/kafka)
* aggregator: one (or more) machine will have a sybil binary and will be used to coordinate queries (across the leaf machines) and stitch results back together
* UI: one machine will have snorkel install with msybil binary. this machine connects to the master aggregator to run sybil queries. might be the same as the master aggregator machine at first. the UI can not be installed on multiple machines at the moment. 

i made some guesses around different type of machine setups here: http://logv.org/distributed-notes.html#example-cluster-setups

**Comment by prologic** [comment]
This sounds initially good to me. Is there support coming for multiple
backends too at some point? Like Presto?

cheers
James


James Mills / prologic

E: prologic@shortcircuit.net.au
W: prologic.shortcircuit.net.au

**Comment by okayzed** [comment]
to understand what you are asking: you want to be able to show multiple datasets from different sources in the snorkel UI? (f.e. we might have some datasets coming from presto and other datasets coming from sybil or mysql tables)

currently, i've only ever been able to support one backend at a time in snorkel but multiple simultaneous backends has been a requested use case in the past. i haven't thought too much about how to accomplish it, but it could be doable. (let's open an issue for it)

my main blocker with presto is that i don't have a presto setup to develop against (and i forget if presto is schemaless, maybe it has per block schemas), but i built a postgres and sqlite adapter that should be easy to copy and modify to work on top of presto, because those adapters build the queries as straight forward SQL. if i'm given a set of presto tables to run queries against (that has a time column in seconds), i can likely add a presto adapter.

**Comment by prologic** [comment]
Yes :) This is something I *think* is worth building (at least allowing for
it).

Some backends to consider supporting:

* sybil
* *sql (postgresql, msqyl, mario)
* textql (https://github.com/dinedal/textql)
* presto

cheers
James


James Mills / prologic

E: prologic@shortcircuit.net.au
W: prologic.shortcircuit.net.au

File: snorkel/config/docker.js [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+    db: "snorkel"
+  },
+  no_api_auth: Boolean(process.env.NO_API_AUTH), // require user authentication to use API endpoint
+  google_auth: {
+    enabled: Boolean(process.env.GOOGLE_AUTH_ENABLED),
+    require_domain: process.env.GPLUS_DOMAIN,
+  },
+  upload_csv: Boolean(process.env.UPLOAD_CSV),
+
+  show_tour: Boolean(process.env.show_tour),
```
Comment (okayzed) [inline snorkel/config/docker.js]: this should be process.env.SHOW_TOUR

**Comment by okayzed** [comment]
@prologic: the mysbil binary has dependency on sybil binary, so i'm fine with including this dependency for now. the plan is to move it out, though - maybe we can all sync up on this offline (on fb chats or IRC)

@tmc: awesome! i tested it out - it works for me. i had to move to root of snorkel/ directory, then create data_dir/config/ and copy the docker.js and users.htpasswd into it, then i was able to log in and start browsing datasets it created. (maybe this can be automated) - anything else left?

**Comment by tmc** [comment]
@prologic given the execution model of sybil it's not trivial to run it in a separate container. I don't think the downsides outweigh the convenience this provides. Distribution and scale will come when sybil can be accessed over a network boundary (msybil).

@okayzed you shouldn't need to copy the docker.js config into data_dir. In terms of additions perhaps having a make target that sets up a demo account or extending the add_user script to support writing in other locations would provide a nicer initial setup experience.

File: snorkel/config/docker.js (hunk 1 above)
Comment (tmc) [reply snorkel/config/docker.js]: Whoops. Thanks. Fixed.

**Comment by okayzed** [comment]
if i don't copy docker.js into `${ROOT}/data_dir/config/`, this is what i see at the beginning of the snorkel run:

```
WARNING: Running server without a $RELEASE - assets will be served in development mode!
CONFIG: LOADING CONFIG DIR FROM /snorkel
CONFIG: Using config in config/config
CONFIG: Adding overrides from config/override.js
CONFIG: Using ENV overrides in config/docker
CONFIG: LOADING CONFIG DIR FROM /var/data_dir
CONFIG: Main config error:Error: Cannot find module '/var/data_dir/config/config'
CONFIG: Couldn't load overrides in config/override.js:Error: Cannot find module '/var/data_dir/config/override'
CONFIG: Couldn't load env conf from config/docker:Error: Cannot find module '/var/data_dir/config/docker'
```

these are small things that can be taken care of later, though - up to you on how far you want to push this patch.

**Comment by prologic** [comment]
@tmc 

> @prologic given the execution model of sybil it's not trivial to run it in a separate container. I don't think the downsides outweigh the convenience this provides. Distribution and scale will come when sybil can be accessed over a network boundary (such as with msybil).

I don't understand. How is snorkel communicating with the sybil backend now? Surely it **must** be doing so over some kind of API (*impliying a network boundary*)

TBH though I haven't looked at the codebase too closely (yet); but I want to use sybil/snorkel in my infra.

**Comment by okayzed** [comment]
@prologic: sybil is a serverless binary that can ingest from stdin and query samples that are on disk. when the ingestion or query finishes, the binary exits. snorkel calls directly into this binary to ingest or execute queries on local data. when doing remote queries, snorkel uses msybil (a python script) to call out to sybil on other machines. the plan is to evolve into a more scalable form (as you were mentioning). 

@tmc will likely be creating a grpc server to sit on top of sybil, s.t. sybil does appear like a regular backend service. then we will be able to arbitrarily add leaf nodes to the sybil swarm and execute queries from the main snorkel server that fan out.

**Comment by tmc** [comment]
@okayzed those warnings are displayed but the config is still read from config/ -- is it better to leave config_dir untouched? I was basing that off of the snapcraft config.

Maybe it's better to not use the label 'docker' since the important thing the docker config is doing is just pulling values from the environment. Is calling this config 'env' or 'environment' too generic/confusing?

**Comment by okayzed** [comment]
@tmc: config_dir/ is so that someone can configure the image externally, otherwise the config will be hardcoded to what was set at packaging time. up to you on what you want to use for the label - env is a tiny bit generic, but it's fine if you copy it over to data_dir/config/env.js or data_dir/config/config.js from snorkel/config/docker.js when packaging. that way, people actually modifying the config will not be confused by the name.

**Comment by tmc** [comment]
My intent was that data_dir is provided as a volume by the user which implies it's not something that could have files put into it at packaging time. Perhaps a good solution is to improve the error message to make it seem like less of an error but more like it's another path that is being checked for configuration?

As long as configuration is declarative in nature I think it should be provided via the environment vs a configuration file.

**Comment by okayzed** [comment]
@tmc: ah, i see - that makes sense re: the not being able to package the config into the volume at build time. how it is now is fine then. you can likely remove the config_dir setting from the config file and those lines will go away.

**Comment by okayzed** [comment]
i guess that's all that's left for accept? (removing config_dir line) and maybe we'll add some script to setup the data_dir/ volume later

**Comment by tmc** [comment]
@okayzed ok, pushed with config_dir removal and I pushed a branch here with the docker->env change -- you pick to include it or now: https://github.com/tmc/snorkel/commit/09d6c8775c1aeecef5159e78dac7c95aa362fe3a

**Comment by okayzed** [comment]
merged in 3c8550d19d71056802a616f27432d5fd232a97b5, i made some slight changes (added config_dir back in via process.env.CONFIG_DIR and npm install sqlite3 --build-from-source to fix #31)

**Pull request closed without merging**